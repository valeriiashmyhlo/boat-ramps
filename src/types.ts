import mapboxgl from "mapbox-gl";

export interface Feature extends mapboxgl.MapboxGeoJSONFeature {
  properties: {
    material: string;
    area_: number;
  };
}

export type Optional<T> = T | null;

export type SizeCounts = { [key: string]: { range: [number, number]; value: number } };

export type MaterialCounts = { [key: string]: number };
