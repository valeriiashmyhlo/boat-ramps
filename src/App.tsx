import "mapbox-gl/dist/mapbox-gl.css";

import { Filters, Optional, Range } from "./types";
import { QueryParamConfig, StringParam, useQueryParam } from "use-query-params";
import React, { useEffect, useMemo, useState } from "react";

import { ChartSection } from "./components/chartSection";
import { LngLatBounds } from "mapbox-gl";
import { Map } from "./components/mapSection";
import { fetchFeatures } from "./api";
import { filteredDataSelector } from "./app.selector";
import { setFeatures } from "./actions";
import { style } from "typestyle";
import { useDispatch } from "react-redux";
import { useTypedSelector } from "./reducers/features.reducer";

const app = style({
  display: "flex",
  textAlign: "center",
});

export const RangeParam: QueryParamConfig<Optional<Range>> = {
  encode: (val) => val && `${val[0]},${val[1]}`,
  decode: (input) => {
    if (input === undefined || input === null || input instanceof Array) {
      return null;
    }
    const a = input.split(",");
    if (a.length !== 2) {
      return null;
    }
    const range = a.map(Number);
    if (range.some(isNaN)) {
      return null;
    }
    return [range[0], range[1]];
  },
};

export const useFilters = () => {
  const [material, setMaterial] = useQueryParam("material", StringParam);
  const [size, setSize] = useQueryParam("range", RangeParam);
  const [bounds, setBounds] = useState<Optional<LngLatBounds>>(null);
  const filters: Filters = useMemo(
    () => ({
      size,
      material,
      bounds,
    }),
    [size, material, bounds]
  );

  return [filters, { setMaterial, setSize, setBounds }] as const;
};

type R<T> = {
  data: Optional<T>;
  isLoading: boolean;
  error: Optional<Error>;
};

function useFetchApi<T>(call: () => Promise<T>): R<T> {
  const [result, setResult] = useState<R<T>>({ data: null, isLoading: false, error: null });

  useEffect(() => {
    (async () => {
      setResult((res) => ({ ...res, isLoading: true }));
      try {
        const data = await call();
        setResult((res) => ({ ...res, data: data, error: null }));
      } catch (e) {
        setResult((res) => ({ ...res, error: e }));
      } finally {
        setResult((res) => ({ ...res, isLoading: false }));
      }
    })();
  }, [call]);

  return result;
}

const App = () => {
  const dispatch = useDispatch();
  const [filters, { setBounds, setMaterial, setSize }] = useFilters();
  const data = useTypedSelector((s) => filteredDataSelector(s, filters));

  useEffect(() => {
    fetchFeatures().then((features) => dispatch(setFeatures(features)));
  }, [dispatch]);

  const result = useFetchApi(fetchFeatures);

  return (
    <div className={app}>
      <Map data={data.features} onBoundsChange={setBounds} />
      <ChartSection
        setFilterMaterial={setMaterial}
        setFilterSize={setSize}
        materialsCounts={data.materialCount}
        sizeCounts={data.sizeCount}
        selectedRange={filters.size}
      />
    </div>
  );
};

export default App;
