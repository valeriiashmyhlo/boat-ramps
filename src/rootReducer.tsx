import { combineReducers } from "redux";
import { featuresReducer } from "./reducers/features.reducer";

export const rootReducer = combineReducers({
  features: featuresReducer,
});

export type RootState = ReturnType<typeof rootReducer>;
