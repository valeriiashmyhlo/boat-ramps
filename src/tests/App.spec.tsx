import App, { RangeParam, useFilters } from "../App";
import { NumberParam, QueryParamProvider, useQueryParam } from "use-query-params";
import { RootState, rootReducer } from "../rootReducer";
import { Route, BrowserRouter as Router } from "react-router-dom";
import { useDispatch, useSelector } from "react-redux";

import { Feature } from "../types";
import { Provider } from "react-redux";
import React from "react";
import { Store } from "../store";
import { createStore } from "redux";
import { render as rtlRender } from "@testing-library/react";

const f1: Feature = {
  type: "Feature",
  geometry: {
    type: "MultiPolygon",
    coordinates: [[[[153, -27]]]],
  },
  properties: {
    material: "Concrete",

    area_: 40.4,
  },
};
const f2: Feature = {
  type: "Feature",
  geometry: {
    type: "MultiPolygon",
    coordinates: [[[[111, 222]]]],
  },
  properties: {
    material: "Gravel",

    area_: 184.4,
  },
};
const f3: Feature = {
  type: "Feature",
  geometry: {
    type: "MultiPolygon",
    coordinates: [[[[111, 222]]]],
  },
  properties: {
    material: "Other",

    area_: 210.4,
  },
};

function render(
  ui: JSX.Element,
  {
    initialState,
    store = createStore(rootReducer, initialState),
    ...renderOptions
  }: { initialState: RootState; store: Store }
) {
  const Wrapper: React.FC<{}> = ({ children }) => <Provider store={store}>{children}</Provider>;
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

test("renders App component", () => {
  const initialState = { features: { features: [f1, f2, f3] } };
  const { container } = render(
    <Router>
      <QueryParamProvider ReactRouterRoute={Route}>
        <App />
      </QueryParamProvider>
    </Router>,
    {
      initialState,
      store: createStore(rootReducer, initialState),
    }
  );
  expect(container.firstChild).toMatchSnapshot();
});

describe("RangeParam", () => {
  it("encode should correctly convert range to string", () => {
    expect(RangeParam.encode([0, 50])).toBe("0,50");
    expect(RangeParam.encode([5, 5])).toBe("5,5");
    expect(RangeParam.encode(null)).toBe(null);
  });
  it("decode should correctly convert string to range", () => {
    expect(RangeParam.decode("5,50")).toEqual([5, 50]);
    expect(RangeParam.decode("0,0")).toEqual([0, 0]);
    expect(RangeParam.decode(null)).toEqual(null);
    expect(RangeParam.decode(undefined)).toEqual(null);
    expect(RangeParam.decode(["1", "2"])).toEqual(null);
  });
});
