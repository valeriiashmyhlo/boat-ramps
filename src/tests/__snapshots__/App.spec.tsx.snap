// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders App component 1`] = `
<div
  class="fjm7llx"
>
  <div
    style="position: relative; width: 0px; height: 0px; cursor: grab; outline: none;"
    tabindex="0"
  >
    <div
      style="position: relative; width: 100%; height: 100%;"
    >
      <div
        class=""
        style="position: absolute; width: 100%; height: 100%; overflow: hidden; visibility: inherit;"
      />
      <div
        style="overflow: visible; height: 0px; width: 0px;"
      />
      <div
        class="resize-triggers"
      >
        <div
          class="expand-trigger"
        >
          <div
            style="width: 1px; height: 1px;"
          />
        </div>
        <div
          class="contract-trigger"
        />
      </div>
    </div>
  </div>
  <div
    class="chartContainer_f1rfmtj7"
  >
    <div
      class="mb20_f4nvz1l"
    >
      Drag the map to populate results
    </div>
    <button
      type="button"
    >
      Reset Materials
    </button>
    <div
      class="rv-xy-plot barChartStyles_f1k0vybx"
      style="width: 550px; height: 300px;"
    >
      <svg
        class="rv-xy-plot__inner"
        height="300"
        width="550"
      >
        <g
          class="rv-xy-plot__axis rv-xy-plot__axis--horizontal "
          transform="translate(115,260)"
        >
          <line
            class="rv-xy-plot__axis__line"
            x1="0"
            x2="425"
            y1="0"
            y2="0"
          />
          <g
            class="rv-xy-plot__axis__ticks"
            transform="translate(0, 0)"
          >
            <g
              class="rv-xy-plot__axis__tick"
              transform="translate(0, 0)"
            >
              <line
                class="rv-xy-plot__axis__tick__line"
                x1="0"
                x2="0"
                y1="-6"
                y2="6"
              />
              <text
                class="rv-xy-plot__axis__tick__text"
                dy="0.72em"
                style="stroke: none; fill: #6b6b76; font-size: 12.5px;"
                text-anchor="middle"
                transform="translate(0, 14)"
              >
                0.0
              </text>
            </g>
            <g
              class="rv-xy-plot__axis__tick"
              transform="translate(42.5, 0)"
            >
              <line
                class="rv-xy-plot__axis__tick__line"
                x1="0"
                x2="0"
                y1="-6"
                y2="6"
              />
              <text
                class="rv-xy-plot__axis__tick__text"
                dy="0.72em"
                style="stroke: none; fill: #6b6b76; font-size: 12.5px;"
                text-anchor="middle"
                transform="translate(0, 14)"
              >
                0.1
              </text>
            </g>
            <g
              class="rv-xy-plot__axis__tick"
              transform="translate(85, 0)"
            >
              <line
                class="rv-xy-plot__axis__tick__line"
                x1="0"
                x2="0"
                y1="-6"
                y2="6"
              />
              <text
                class="rv-xy-plot__axis__tick__text"
                dy="0.72em"
                style="stroke: none; fill: #6b6b76; font-size: 12.5px;"
                text-anchor="middle"
                transform="translate(0, 14)"
              >
                0.2
              </text>
            </g>
            <g
              class="rv-xy-plot__axis__tick"
              transform="translate(127.5, 0)"
            >
              <line
                class="rv-xy-plot__axis__tick__line"
                x1="0"
                x2="0"
                y1="-6"
                y2="6"
              />
              <text
                class="rv-xy-plot__axis__tick__text"
                dy="0.72em"
                style="stroke: none; fill: #6b6b76; font-size: 12.5px;"
                text-anchor="middle"
                transform="translate(0, 14)"
              >
                0.3
              </text>
            </g>
            <g
              class="rv-xy-plot__axis__tick"
              transform="translate(170, 0)"
            >
              <line
                class="rv-xy-plot__axis__tick__line"
                x1="0"
                x2="0"
                y1="-6"
                y2="6"
              />
              <text
                class="rv-xy-plot__axis__tick__text"
                dy="0.72em"
                style="stroke: none; fill: #6b6b76; font-size: 12.5px;"
                text-anchor="middle"
                transform="translate(0, 14)"
              >
                0.4
              </text>
            </g>
            <g
              class="rv-xy-plot__axis__tick"
              transform="translate(212.5, 0)"
            >
              <line
                class="rv-xy-plot__axis__tick__line"
                x1="0"
                x2="0"
                y1="-6"
                y2="6"
              />
              <text
                class="rv-xy-plot__axis__tick__text"
                dy="0.72em"
                style="stroke: none; fill: #6b6b76; font-size: 12.5px;"
                text-anchor="middle"
                transform="translate(0, 14)"
              >
                0.5
              </text>
            </g>
            <g
              class="rv-xy-plot__axis__tick"
              transform="translate(255, 0)"
            >
              <line
                class="rv-xy-plot__axis__tick__line"
                x1="0"
                x2="0"
                y1="-6"
                y2="6"
              />
              <text
                class="rv-xy-plot__axis__tick__text"
                dy="0.72em"
                style="stroke: none; fill: #6b6b76; font-size: 12.5px;"
                text-anchor="middle"
                transform="translate(0, 14)"
              >
                0.6
              </text>
            </g>
            <g
              class="rv-xy-plot__axis__tick"
              transform="translate(297.5, 0)"
            >
              <line
                class="rv-xy-plot__axis__tick__line"
                x1="0"
                x2="0"
                y1="-6"
                y2="6"
              />
              <text
                class="rv-xy-plot__axis__tick__text"
                dy="0.72em"
                style="stroke: none; fill: #6b6b76; font-size: 12.5px;"
                text-anchor="middle"
                transform="translate(0, 14)"
              >
                0.7
              </text>
            </g>
            <g
              class="rv-xy-plot__axis__tick"
              transform="translate(340, 0)"
            >
              <line
                class="rv-xy-plot__axis__tick__line"
                x1="0"
                x2="0"
                y1="-6"
                y2="6"
              />
              <text
                class="rv-xy-plot__axis__tick__text"
                dy="0.72em"
                style="stroke: none; fill: #6b6b76; font-size: 12.5px;"
                text-anchor="middle"
                transform="translate(0, 14)"
              >
                0.8
              </text>
            </g>
            <g
              class="rv-xy-plot__axis__tick"
              transform="translate(382.5, 0)"
            >
              <line
                class="rv-xy-plot__axis__tick__line"
                x1="0"
                x2="0"
                y1="-6"
                y2="6"
              />
              <text
                class="rv-xy-plot__axis__tick__text"
                dy="0.72em"
                style="stroke: none; fill: #6b6b76; font-size: 12.5px;"
                text-anchor="middle"
                transform="translate(0, 14)"
              >
                0.9
              </text>
            </g>
            <g
              class="rv-xy-plot__axis__tick"
              transform="translate(425, 0)"
            >
              <line
                class="rv-xy-plot__axis__tick__line"
                x1="0"
                x2="0"
                y1="-6"
                y2="6"
              />
              <text
                class="rv-xy-plot__axis__tick__text"
                dy="0.72em"
                style="stroke: none; fill: #6b6b76; font-size: 12.5px;"
                text-anchor="middle"
                transform="translate(0, 14)"
              >
                1.0
              </text>
            </g>
          </g>
        </g>
        <g
          class="rv-xy-plot__axis rv-xy-plot__axis--vertical "
          transform="translate(9,10)"
        >
          <line
            class="rv-xy-plot__axis__line"
            x1="115"
            x2="115"
            y1="0"
            y2="250"
          />
          <g
            class="rv-xy-plot__axis__ticks"
            transform="translate(115, 0)"
          >
            <g
              class="rv-xy-plot__axis__tick"
              transform="translate(0, 208.33333333333331)"
            >
              <line
                class="rv-xy-plot__axis__tick__line"
                x1="6"
                x2="-6"
                y1="0"
                y2="0"
              />
              <text
                class="rv-xy-plot__axis__tick__text"
                dy="0.32em"
                style="stroke: none; fill: #6b6b76; font-size: 12.5px;"
                text-anchor="end"
                transform="translate(-14, 0)"
              >
                Concrete
              </text>
            </g>
            <g
              class="rv-xy-plot__axis__tick"
              transform="translate(0, 125)"
            >
              <line
                class="rv-xy-plot__axis__tick__line"
                x1="6"
                x2="-6"
                y1="0"
                y2="0"
              />
              <text
                class="rv-xy-plot__axis__tick__text"
                dy="0.32em"
                style="stroke: none; fill: #6b6b76; font-size: 12.5px;"
                text-anchor="end"
                transform="translate(-14, 0)"
              >
                Gravel
              </text>
            </g>
            <g
              class="rv-xy-plot__axis__tick"
              transform="translate(0, 41.66666666666667)"
            >
              <line
                class="rv-xy-plot__axis__tick__line"
                x1="6"
                x2="-6"
                y1="0"
                y2="0"
              />
              <text
                class="rv-xy-plot__axis__tick__text"
                dy="0.32em"
                style="stroke: none; fill: #6b6b76; font-size: 12.5px;"
                text-anchor="end"
                transform="translate(-14, 0)"
              >
                Other
              </text>
            </g>
          </g>
        </g>
        <g
          class="rv-xy-plot__series rv-xy-plot__series--bar "
          transform="translate(115,10)"
        >
          <rect
            height="62.499999999999986"
            style="opacity: 0.3; stroke: grey; fill: grey;"
            width="425"
            x="0"
            y="177.08333333333331"
          />
          <rect
            height="62.499999999999986"
            style="opacity: 0.3; stroke: grey; fill: grey;"
            width="425"
            x="0"
            y="93.75"
          />
          <rect
            height="62.499999999999986"
            style="opacity: 0.3; stroke: grey; fill: grey;"
            width="425"
            x="0"
            y="10.416666666666679"
          />
        </g>
      </svg>
    </div>
    <button
      type="button"
    >
      Reset Range
    </button>
    <div
      class="RadialChartWrap_f1c7x5pa"
    >
      <div
        class="rv-xy-plot radialChartStyles_f1k2oovd rv-radial-chart"
        style="width: 300px; height: 300px;"
      >
        <svg
          class="rv-xy-plot__inner"
          height="300"
          width="300"
        >
          <g
            class="rv-xy-plot__series--arc__animation-wrapper"
          >
            <g
              class="rv-xy-plot__series rv-xy-plot__series--arc radialChartStyles_f1k2oovd"
              opacity="1"
              pointer-events="none"
              transform="translate(150,150)"
            >
              <path
                class="rv-xy-plot__series rv-xy-plot__series--arc-path rv-radial-chart__series--pie__slice "
                d="M-130.74545369987374,73.52296469004888A150,150,0,0,1,-1.6999716668084133,-149.99036667843723L-1.699971666808315,-79.98193606266385A80,80,0,0,0,-70.11637430753377,38.51874938216216Z"
                style="opacity: 1; stroke: 0; fill: rgb(17, 147, 153); stroke-width: 0;"
              />
              <path
                class="rv-xy-plot__series rv-xy-plot__series--arc-path rv-radial-chart__series--pie__slice "
                d="M129.04548203306544,76.46740198838837A150,150,0,0,1,-129.0454820330654,76.46740198838845L-68.41640264072544,41.46318668050175A80,80,0,0,0,68.41640264072547,41.46318668050171Z"
                style="opacity: 1; stroke: 1px; fill: rgb(121, 199, 227); stroke-width: 0;"
              />
              <path
                class="rv-xy-plot__series rv-xy-plot__series--arc-path rv-radial-chart__series--pie__slice "
                d="M1.6999716668083285,-149.99036667843723A150,150,0,0,1,130.74545369987376,73.5229646900488L70.11637430753379,38.51874938216212A80,80,0,0,0,1.6999716668083407,-79.98193606266385Z"
                style="opacity: 1; stroke: 2px; fill: rgb(26, 49, 119); stroke-width: 0;"
              />
            </g>
            <g
              class="rv-xy-plot__series rv-xy-plot__series--arc radialChartStyles_f1k2oovd"
              opacity="0"
              pointer-events="all"
              transform="translate(150,150)"
            >
              <path
                class="rv-xy-plot__series rv-xy-plot__series--arc-path rv-radial-chart__series--pie__slice "
                d="M-130.74545369987374,73.52296469004888A150,150,0,0,1,-1.6999716668084133,-149.99036667843723L-1.699971666808315,-79.98193606266385A80,80,0,0,0,-70.11637430753377,38.51874938216216Z"
                style="opacity: 1; stroke: red; fill: #119399; stroke-width: 0;"
              />
              <path
                class="rv-xy-plot__series rv-xy-plot__series--arc-path rv-radial-chart__series--pie__slice "
                d="M129.04548203306544,76.46740198838837A150,150,0,0,1,-129.0454820330654,76.46740198838845L-68.41640264072544,41.46318668050175A80,80,0,0,0,68.41640264072547,41.46318668050171Z"
                style="opacity: 1; stroke: red; fill: #79c7e3; stroke-width: 0;"
              />
              <path
                class="rv-xy-plot__series rv-xy-plot__series--arc-path rv-radial-chart__series--pie__slice "
                d="M1.6999716668083285,-149.99036667843723A150,150,0,0,1,130.74545369987376,73.5229646900488L70.11637430753379,38.51874938216212A80,80,0,0,0,1.6999716668083407,-79.98193606266385Z"
                style="opacity: 1; stroke: red; fill: #1a3177; stroke-width: 0;"
              />
            </g>
          </g>
        </svg>
        <div
          class="radialChartLabel_f45ezqg"
          datatest-id="radialChartLabel"
        >
          null
           sizes shown
        </div>
      </div>
      <div>
        <div
          class="chartAnnotation_f1gqgsum"
        >
          <span
            class="annotation_ftgu0ol"
            style="background-color: rgb(26, 49, 119);"
          />
          -
           
          [0,50]
        </div>
        <div
          class="chartAnnotation_f1gqgsum"
        >
          <span
            class="annotation_ftgu0ol"
            style="background-color: rgb(121, 199, 227);"
          />
          -
           
          [50,200]
        </div>
        <div
          class="chartAnnotation_f1gqgsum"
        >
          <span
            class="annotation_ftgu0ol"
            style="background-color: rgb(17, 147, 153);"
          />
          -
           
          [200,526]
        </div>
      </div>
    </div>
  </div>
</div>
`;
